File: packages/sdk-node/src/bridges/telegram.ts
import type { Bridge, BridgeMessageResult } from './types';
import type { PocketPing } from '../pocketping';
import type { Session, Message } from '../types';
import { PocketPingSetupError, SETUP_GUIDES } from '../errors';

/**
 * Telegram API response types
 */
interface TelegramResponse {
  ok: boolean;
  description?: string;
  result?: {
    message_id?: number;
    [key: string]: unknown;
  };
}

/**
 * Options for TelegramBridge
 */
export interface TelegramBridgeOptions {
  /** Parse mode for message formatting */
  parseMode?: 'HTML' | 'Markdown';
  /** Disable notification sound */
  disableNotification?: boolean;
}

/**
 * Telegram Bridge for PocketPing.
 * Sends visitor messages to a Telegram chat using the Bot API.
 *
 * @example
 * ```ts
 * const telegram = new TelegramBridge(
 *   'BOT_TOKEN',
 *   '-1001234567890',
 *   { parseMode: 'HTML' }
 * );
 * const pocketping = new PocketPing({ bridges: [telegram] });
 * ```
 */
export class TelegramBridge implements Bridge {
  readonly name = 'telegram';

  private pocketping?: PocketPing;
  private readonly botToken: string;
  private readonly chatId: string | number;
  private readonly parseMode: 'HTML' | 'Markdown';
  private readonly disableNotification: boolean;
  private readonly baseUrl: string;

  constructor(
    botToken: string,
    chatId: string | number,
    options: TelegramBridgeOptions = {}
  ) {
    if (!botToken) {
      throw new PocketPingSetupError({
        bridge: 'Telegram',
        missing: 'botToken',
        guide: SETUP_GUIDES.telegram.botToken,
      });
    }

    if (!chatId) {
      throw new PocketPingSetupError({
        bridge: 'Telegram',
        missing: 'chatId',
        guide: SETUP_GUIDES.telegram.chatId,
      });
    }

    this.botToken = botToken;
    this.chatId = chatId;
    this.parseMode = options.parseMode ?? 'HTML';
    this.disableNotification = options.disableNotification ?? false;
    this.baseUrl = `https://api.telegram.org/bot${botToken}`;
  }

  /**
   * Initialize the bridge (optional setup)
   */
  async init(pocketping: PocketPing): Promise<void> {
    this.pocketping = pocketping;
    // Optionally verify bot token by calling getMe
    try {
      const response = await fetch(`${this.baseUrl}/getMe`);
      const data = (await response.json()) as TelegramResponse;
      if (!data.ok) {
        console.error('[TelegramBridge] Invalid bot token:', data.description);
      }
    } catch (error) {
      console.error('[TelegramBridge] Failed to verify bot token:', error);
    }
  }

  /**
   * Called when a new chat session is created
   */
  async onNewSession(session: Session): Promise<void> {
    const url = session.metadata?.url || 'Unknown page';
    const text = this.formatNewSession(session.visitorId, url);

    try {
      await this.sendMessage(text);
    } catch (error) {
      console.error('[TelegramBridge] Failed to send new session notification:', error);
    }
  }

  /**
   * Called when a visitor sends a message.
   * Returns the Telegram message ID for edit/delete sync.
   */
  async onVisitorMessage(
    message: Message,
    session: Session
  ): Promise<BridgeMessageResult> {
    const text = this.formatVisitorMessage(session.visitorId, message.content);
    let replyToMessageId: number | undefined;

    if (message.replyTo) {
      const storage = this.pocketping?.getStorage();
      if (storage?.getBridgeMessageIds) {
        const ids = await storage.getBridgeMessageIds(message.replyTo);
        if (ids?.telegramMessageId) {
          replyToMessageId = ids.telegramMessageId;
        }
      }
    }

    try {
      const messageId = await this.sendMessage(text, replyToMessageId);
      return { messageId };
    } catch (error) {
      console.error('[TelegramBridge] Failed to send visitor message:', error);
      return {};
    }
  }

  /**
   * Called when an operator sends a message (for cross-bridge sync)
   */
  async onOperatorMessage(
    message: Message,
    _session: Session,
    sourceBridge?: string,
    operatorName?: string
  ): Promise<void> {
    // Don't echo messages that originated from Telegram
    if (sourceBridge === 'telegram') {
      return;
    }

    const name = operatorName || 'Operator';
    const text = this.parseMode === 'HTML'
      ? `<b>${this.escapeHtml(name)}:</b>\n${this.escapeHtml(message.content)}`
      : `*${this.escapeMarkdown(name)}:*\n${this.escapeMarkdown(message.content)}`;

    try {
      await this.sendMessage(text);
    } catch (error) {
      console.error('[TelegramBridge] Failed to send operator message:', error);
    }
  }

  /**
   * Called when visitor starts/stops typing
   */
  async onTyping(sessionId: string, isTyping: boolean): Promise<void> {
    if (!isTyping) return;

    try {
      await this.sendChatAction('typing');
    } catch (error) {
      console.error('[TelegramBridge] Failed to send typing action:', error);
    }
  }

  /**
   * Called when a visitor edits their message.
   * @returns true if edit succeeded, false otherwise
   */
  async onMessageEdit(
    _messageId: string,
    newContent: string,
    bridgeMessageId: string | number
  ): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/editMessageText`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          chat_id: this.chatId,
          message_id: bridgeMessageId,
          text: `${newContent}\n\n<i>(edited)</i>`,
          parse_mode: this.parseMode,
        }),
      });

      const data = (await response.json()) as TelegramResponse;
      if (!data.ok) {
        console.error('[TelegramBridge] Edit failed:', data.description);
        return false;
      }
      return true;
    } catch (error) {
      console.error('[TelegramBridge] Failed to edit message:', error);
      return false;
    }
  }

  /**
   * Called when a visitor deletes their message.
   * @returns true if delete succeeded, false otherwise
   */
  async onMessageDelete(
    _messageId: string,
    bridgeMessageId: string | number
  ): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/deleteMessage`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          chat_id: this.chatId,
          message_id: bridgeMessageId,
        }),
      });

      const data = (await response.json()) as TelegramResponse;
      if (!data.ok) {
        // Message might already be deleted or too old
        console.error('[TelegramBridge] Delete failed:', data.description);
        return false;
      }
      return true;
    } catch (error) {
      console.error('[TelegramBridge] Failed to delete message:', error);
      return false;
    }
  }

  /**
   * Called when a custom event is triggered from the widget
   */
  async onCustomEvent(
    event: { name: string; data?: Record<string, unknown> },
    session: Session
  ): Promise<void> {
    const dataStr = event.data ? JSON.stringify(event.data, null, 2) : '';
    const text = this.parseMode === 'HTML'
      ? `<b>Custom Event:</b> ${this.escapeHtml(event.name)}\n<b>Visitor:</b> ${this.escapeHtml(session.visitorId)}${dataStr ? `\n<pre>${this.escapeHtml(dataStr)}</pre>` : ''}`
      : `*Custom Event:* ${this.escapeMarkdown(event.name)}\n*Visitor:* ${this.escapeMarkdown(session.visitorId)}${dataStr ? `\n\`\`\`\n${dataStr}\n\`\`\`` : ''}`;

    try {
      await this.sendMessage(text);
    } catch (error) {
      console.error('[TelegramBridge] Failed to send custom event:', error);
    }
  }

  /**
   * Called when a user identifies themselves via PocketPing.identify()
   */
  async onIdentityUpdate(session: Session): Promise<void> {
    if (!session.identity) return;

    const identity = session.identity;
    let text: string;

    if (this.parseMode === 'HTML') {
      text = `<b>User Identified</b>\n` +
        `<b>ID:</b> ${this.escapeHtml(identity.id)}\n` +
        (identity.name ? `<b>Name:</b> ${this.escapeHtml(identity.name)}\n` : '') +
        (identity.email ? `<b>Email:</b> ${this.escapeHtml(identity.email)}\n` : '') +
        (session.userPhone ? `<b>Phone:</b> ${this.escapeHtml(session.userPhone)}` : '');
    } else {
      text = `*User Identified*\n` +
        `*ID:* ${this.escapeMarkdown(identity.id)}\n` +
        (identity.name ? `*Name:* ${this.escapeMarkdown(identity.name)}\n` : '') +
        (identity.email ? `*Email:* ${this.escapeMarkdown(identity.email)}\n` : '') +
        (session.userPhone ? `*Phone:* ${this.escapeMarkdown(session.userPhone)}` : '');
    }

    try {
      await this.sendMessage(text.trim());
    } catch (error) {
      console.error('[TelegramBridge] Failed to send identity update:', error);
    }
  }

  // ─────────────────────────────────────────────────────────────────
  // Private helper methods
  // ─────────────────────────────────────────────────────────────────

  /**
   * Send a message to the Telegram chat
   */
  private async sendMessage(
    text: string,
    replyToMessageId?: number
  ): Promise<number | undefined> {
    const response = await fetch(`${this.baseUrl}/sendMessage`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        chat_id: this.chatId,
        text,
        parse_mode: this.parseMode,
        disable_notification: this.disableNotification,
        ...(replyToMessageId ? { reply_to_message_id: replyToMessageId } : {}),
      }),
    });

    const data = (await response.json()) as TelegramResponse;
    if (!data.ok) {
      throw new Error(`Telegram API error: ${data.description}`);
    }
    return data.result?.message_id;
  }

  /**
   * Send a chat action (e.g., "typing")
   */
  private async sendChatAction(action: string): Promise<void> {
    await fetch(`${this.baseUrl}/sendChatAction`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        chat_id: this.chatId,
        action,
      }),
    });
  }

  /**
   * Format new session notification
   */
  private formatNewSession(visitorId: string, url: string): string {
    if (this.parseMode === 'HTML') {
      return `<b>New chat session</b>\n` +
        `<b>Visitor:</b> ${this.escapeHtml(visitorId)}\n` +
        `<b>Page:</b> ${this.escapeHtml(url)}`;
    }
    return `*New chat session*\n` +
      `*Visitor:* ${this.escapeMarkdown(visitorId)}\n` +
      `*Page:* ${this.escapeMarkdown(url)}`;
  }

  /**
   * Format visitor message
   */
  private formatVisitorMessage(visitorId: string, content: string): string {
    if (this.parseMode === 'HTML') {
      return `<b>${this.escapeHtml(visitorId)}:</b>\n${this.escapeHtml(content)}`;
    }
    return `*${this.escapeMarkdown(visitorId)}:*\n${this.escapeMarkdown(content)}`;
  }

  /**
   * Escape HTML special characters
   */
  private escapeHtml(text: string): string {
    return text
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;');
  }

  /**
   * Escape Markdown special characters
   */
  private escapeMarkdown(text: string): string {
    return text.replace(/[_*[\]()~`>#+=|{}.!-]/g, '\\$&');
  }
}

File: bridge-server/internal/bridges/telegram.go
package bridges

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	pocketping "github.com/Ruwad-io/pocketping/sdk-go"
	"github.com/pocketping/bridge-server/internal/config"
	"github.com/pocketping/bridge-server/internal/types"
)

const telegramAPIBase = "https://api.telegram.org/bot"

// TelegramBridge sends notifications to Telegram
type TelegramBridge struct {
	*BaseBridge
	botToken string
	chatID   string
	client   *http.Client
}

// NewTelegramBridge creates a new Telegram bridge with validation
func NewTelegramBridge(cfg *config.TelegramConfig) (*TelegramBridge, error) {
	if err := pocketping.ValidateTelegramConfig(cfg.BotToken, cfg.ChatID); err != nil {
		return nil, err
	}

	return &TelegramBridge{
		BaseBridge: NewBaseBridge("telegram"),
		botToken:   cfg.BotToken,
		chatID:     cfg.ChatID,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}, nil
}

// telegramResponse is the standard Telegram API response
type telegramResponse struct {
	OK          bool            `json:"ok"`
	Result      json.RawMessage `json:"result,omitempty"`
	Description string          `json:"description,omitempty"`
}

// telegramMessageResult is the result of sending a message
type telegramMessageResult struct {
	MessageID int `json:"message_id"`
}

// callAPI makes a request to the Telegram Bot API
func (b *TelegramBridge) callAPI(method string, data map[string]interface{}) (*telegramResponse, error) {
	url := fmt.Sprintf("%s%s/%s", telegramAPIBase, b.botToken, method)

	body, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := b.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result telegramResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, err
	}

	return &result, nil
}

// sendMessage sends a message to the configured chat
func (b *TelegramBridge) sendMessage(text string, replyToMessageID *int) (int, error) {
	data := map[string]interface{}{
		"chat_id":    b.chatID,
		"text":       text,
		"parse_mode": "HTML",
	}
	if replyToMessageID != nil && *replyToMessageID > 0 {
		data["reply_to_message_id"] = *replyToMessageID
	}

	resp, err := b.callAPI("sendMessage", data)
	if err != nil {
		return 0, err
	}

	if !resp.OK {
		log.Printf("[TelegramBridge] API error: %s", resp.Description)
		return 0, fmt.Errorf("telegram API error: %s", resp.Description)
	}

	var msgResult telegramMessageResult
	if err := json.Unmarshal(resp.Result, &msgResult); err != nil {
		return 0, err
	}

	return msgResult.MessageID, nil
}

// OnNewSession announces a new chat session
func (b *TelegramBridge) OnNewSession(session *types.Session) error {
	visitorName := session.VisitorID
	if session.Identity != nil && session.Identity.Name != "" {
		visitorName = session.Identity.Name
	}

	text := fmt.Sprintf("🆕 <b>New chat session</b>\n👤 Visitor: %s", visitorName)

	if session.Metadata != nil {
		if session.Metadata.Country != "" || session.Metadata.City != "" {
			text += fmt.Sprintf("\n🌍 %s, %s", session.Metadata.Country, session.Metadata.City)
		}
		if session.Metadata.URL != "" {
			text += fmt.Sprintf("\n📍 %s", session.Metadata.URL)
		}
	}

	_, err := b.sendMessage(text, nil)
	return err
}

// OnVisitorMessage sends a visitor message to the chat
func (b *TelegramBridge) OnVisitorMessage(message *types.Message, session *types.Session, reply *ReplyContext) (*types.BridgeMessageIDs, error) {
	visitorName := session.VisitorID
	if session.Identity != nil && session.Identity.Name != "" {
		visitorName = session.Identity.Name
	}

	text := fmt.Sprintf("💬 <b>%s</b>:\n%s", visitorName, message.Content)

	if len(message.Attachments) > 0 {
		text += fmt.Sprintf("\n📎 %d attachment(s)", len(message.Attachments))
	}

	var replyToMessageID *int
	if reply != nil && reply.BridgeIDs != nil && reply.BridgeIDs.TelegramMessageID != 0 {
		id := reply.BridgeIDs.TelegramMessageID
		replyToMessageID = &id
	}

	msgID, err := b.sendMessage(text, replyToMessageID)
	if err != nil {
		return nil, err
	}

	return &types.BridgeMessageIDs{TelegramMessageID: msgID}, nil
}

// OnOperatorMessage relays an operator message from another bridge
func (b *TelegramBridge) OnOperatorMessage(message *types.Message, session *types.Session, sourceBridge, operatorName string) error {
	// Don't echo messages from this bridge
	if sourceBridge == "telegram" {
		return nil
	}

	name := operatorName
	if name == "" {
		name = "Operator"
	}

	text := fmt.Sprintf("👤 <b>%s</b> (via %s):\n%s", name, sourceBridge, message.Content)
	_, err := b.sendMessage(text, nil)
	return err
}

// OnTyping sends a typing indicator
func (b *TelegramBridge) OnTyping(sessionID string, isTyping bool) error {
	if !isTyping {
		return nil
	}

	data := map[string]interface{}{
		"chat_id": b.chatID,
		"action":  "typing",
	}

	_, err := b.callAPI("sendChatAction", data)
	return err
}

// OnMessageRead handles read receipts (no-op for Telegram)
func (b *TelegramBridge) OnMessageRead(sessionID string, messageIDs []string, status types.MessageStatus) error {
	return nil
}

// OnCustomEvent sends a custom event notification
func (b *TelegramBridge) OnCustomEvent(event *types.CustomEvent, session *types.Session) error {
	text := fmt.Sprintf("⚡ <b>Event: %s</b>", event.Name)
	if event.Data != nil {
		data, _ := json.Marshal(event.Data)
		text += fmt.Sprintf("\n<code>%s</code>", string(data))
	}
	_, err := b.sendMessage(text, nil)
	return err
}

// OnIdentityUpdate sends an identity update notification
func (b *TelegramBridge) OnIdentityUpdate(session *types.Session) error {
	if session.Identity == nil {
		return nil
	}

	text := fmt.Sprintf("🔑 <b>User identified</b>\nID: %s", session.Identity.ID)
	if session.Identity.Name != "" {
		text += fmt.Sprintf("\nName: %s", session.Identity.Name)
	}
	if session.Identity.Email != "" {
		text += fmt.Sprintf("\nEmail: %s", session.Identity.Email)
	}
	if session.UserPhone != "" {
		text += fmt.Sprintf("\n📱 Phone: %s", session.UserPhone)
	}

	_, err := b.sendMessage(text, nil)
	return err
}

// OnAITakeover sends an AI takeover notification
func (b *TelegramBridge) OnAITakeover(session *types.Session, reason string) error {
	text := fmt.Sprintf("🤖 <b>AI Takeover</b>\nReason: %s", reason)
	_, err := b.sendMessage(text, nil)
	return err
}

// OnVisitorMessageEdited syncs a message edit to Telegram
func (b *TelegramBridge) OnVisitorMessageEdited(sessionID, messageID, content string, bridgeIDs *types.BridgeMessageIDs) (*types.BridgeMessageIDs, error) {
	if bridgeIDs == nil || bridgeIDs.TelegramMessageID == 0 {
		return nil, nil
	}

	data := map[string]interface{}{
		"chat_id":    b.chatID,
		"message_id": bridgeIDs.TelegramMessageID,
		"text":       fmt.Sprintf("✏️ (edited):\n%s", content),
		"parse_mode": "HTML",
	}

	resp, err := b.callAPI("editMessageText", data)
	if err != nil {
		return nil, err
	}

	if !resp.OK {
		log.Printf("[TelegramBridge] Edit failed: %s", resp.Description)
		return nil, nil
	}

	return &types.BridgeMessageIDs{TelegramMessageID: bridgeIDs.TelegramMessageID}, nil
}

// OnVisitorMessageDeleted syncs a message delete to Telegram
func (b *TelegramBridge) OnVisitorMessageDeleted(sessionID, messageID string, bridgeIDs *types.BridgeMessageIDs) error {
	if bridgeIDs == nil || bridgeIDs.TelegramMessageID == 0 {
		return nil
	}

	data := map[string]interface{}{
		"chat_id":    b.chatID,
		"message_id": bridgeIDs.TelegramMessageID,
	}

	resp, err := b.callAPI("deleteMessage", data)
	if err != nil {
		return err
	}

	if !resp.OK {
		log.Printf("[TelegramBridge] Delete failed: %s", resp.Description)
	}

	return nil
}
